### ISSUE TITLE
Add a method to simplify a SpecifierSet

### ISSUE THREAD

**Issue by pfmoore** [pypa/packaging#776]
It would be useful when reasoning about `SpecifierSet` values to have a means of simplifying them. For example, `SpecifierSet("<1.0, <2.0")` is equivalent to `SpecifierSet("<1.0")`. For example, knowing that a complex specifier set is guaranteed to be empty can be useful when resolving requirements.

I have a prototype implementation which currently only merges multiple "<", "<=", ">" and ">=" specifiers, but which could be extended to simplify uses of other specifiers. I'd be willing to develop this into a full implementation and contribute it if there was interest.

One technical issue which would need discussion is around pre-release specifiers. The specifier set `>1.0a1, >2.0` is equivalent to `>2.0`. However, the former "explicitly mentions" a prerelease version, whereas the latter does not. I'm not 100% sure what the packaging library's policy is on whether this matters when checking if a version satisfies a specifier. There are a number of options for this case - simplify anyway, refuse to simplify, or simplify and return a flag saying that an explicit pre-release version was eliminated in the simplification.

**Comment by notatallshaw** [comment]
FYI, I also have a use case for this.

I have an idea to inspect specifiers when resolving requirements. Simplifying specifiers first can reduce complexity, and having unambiguous answers to the questions raised here would save having to make guesses on how they should behave.

**Comment by pfmoore** [comment]
Yes, this was prompted by https://github.com/pypa/packaging/issues/760, because a specifier like `>1.0, <1.0` is empty, but the check suggested in that issue wouldn't identify that. Simplifying and *then* checking would.

At the moment, I'm assuming a separate method, `simplified = spec.simplify()`, but maybe there's a case for maintaining specifier sets in simplified form at all times? The only methods that create specifier sets are the constructor and `__and__`, and it would be easy enough to simplify as a final step in each of those.

**Comment by pfmoore** [comment]
> I'm not 100% sure what the packaging library's policy is on whether this matters when checking if a version satisfies a specifier.

OK, so I tested:

```
>>> SpecifierSet(">=1.0.0a1,>1.1.0").contains("1.3.0a1")
True
>>> SpecifierSet(">1.1.0").contains("1.3.0a1")
False
>>> SpecifierSet(">1.1.0",prereleases=True).contains("1.3.0a1")
True
```

To be correct, it looks like `SpecifierSet(">=1.0.0a1,>1.1.0").contains("1.3.0a1")` would have to simplify to `SpecifierSet(">1.1.0",prereleases=True)`. I'm a little concerned that the default handling of `prereleases` (when it's not specified, or `None`, in the constructor arguments) doesn't behave as I expected:

```
>>> SpecifierSet(">1.1.0a1").prereleases
False
```

I'd have been perfectly happy with either `True` or `None` here - but `False` just seems wrong to me. Can one of the maintainers explain the logic here?

**Comment by pfmoore** [comment]
> Can one of the maintainers explain the logic here?

It appears that the logic is [here](https://github.com/pypa/packaging/blob/main/src/packaging/specifiers.py#L257) - only *inclusive* operators count as "explicitly requesting" a pre-release, so `>1.1.0a1` does not match pre-releases. I can't say that I particularly agree with this interpretation, but it's deliberate and explicit, so I guess I'll need to follow it when simplifying.

**Comment by pfmoore** [comment]
> Yes, this was prompted by https://github.com/pypa/packaging/issues/760, because a specifier like >1.0, <1.0 is empty, but the check suggested in that issue wouldn't identify that. Simplifying and then checking would.

Wait. It appears that I misunderstood what "empty" meant in that issue. `SpecifierSet("")` is empty in the sense that its length is 0. But this means that it contains *every* version (it has no specifiers that limit the allowed versions). What I was thinking was that it was empty because it contained *no* versions.

I can't actually think of a way to write a specifier that is empty in the sense that it contains no versions, that is simpler than `<v,>v` (for any version v). Is that correct?

**Comment by notatallshaw** [comment]
> I can't actually think of a way to write a specifier that is empty in the sense that it contains no versions, that is simpler than `<v,>v` (for any version v). Is that correct?

Correct, I had a think about that in this issue: https://github.com/pypa/packaging/issues/762.

You probably got this misunderstanding from me, as I had that misunderstanding when I wrote https://github.com/pypa/packaging/issues/760. A “simplify” would still be useful, i.e. ">2,>1" gets simplified to ">2".

But what is also needed, at least for my use case, is the ability to ask "for this Specifier Set is it completely contradictory? I.e. there are no possible versions that could match it", this would have to be a different method than "simplify" as the result is not expressible itself as a Specifier Set. Also this method may need to be careful with its return values, it probably needs to be explicit in having a "can not determine" value.

Apologies if this is outside the scope of what you were initially thinking.

**Comment by dstufft** [comment]
> I'd have been perfectly happy with either True or None here - but False just seems wrong to me. Can one of the maintainers explain the logic here?

``SpecifierSet.prereleases`` encapsulates the logic to handle both explicitly specified via kwarg and determined via inspecting the specifier.

I'm pretty sure the answer is just that when I wrote the code 10 years ago, I was trying to prevent `!=1.0a1` from triggering the "should include the prerelease`, and so I included a check not just on what versions were mentioned in the specifier, but whether that specifier logically was "including" that version or whether it was "excluding" that version.

I think this makes general sense, ``!=1.0a1`` is explicitly saying you don't want that version, so it _not_ requesting prereleases, while `==1.0a1` is explicitly saying you _do_ want that version, so it _is_ requesting prereleases.

The `>` and `<` operators kind of sit in a weird place though, `>1.0a1` is _not_ requesting 1.0a1, it's requesting anything of a higher version than that, so it currently does not trigger the "user has requested prereleases" logic, but that creates a weird asymmetry with `>=1.0a1`. To make matters worse, It's not possible to get a prerelease version that is less than `1.0a0.dev0`, so does `<1.0a0.dev0` include or exclude prereleases?

 I don't remember thinking to hard about it, and just did the ones that explicitly included the pre-release that was being mentioned, which meant that `>` and `<` were excluded from being considered triggering the "user has requested a prerelease" logic... but I dont' feel really strongly about that one way or another.

**Comment by pfmoore** [comment]
No, it's *precisely* in scope of what I'm thinking about. And the result *is* expressible as a specifier set (after all, ">0,<0" is precisely such a set) - it's just that there's no immediately obvious "canonical" answer. A simplifier could certainly just pick a value and consider it canonical (and reduce any other provably-unsatisfiable specifier sets to that value).

What's actually far more of an issue is the handling of prerelease versions, which I'm starting to think may actually make this impossible. Consider the following example:

```
>>> Version("0.dev0") in Specifier("<0")
False
>>> Version("0.dev0") in Specifier("<=0.dev0")
True
>>> Version("0.dev0") in SpecifierSet("<0,<=0.dev0")
False
```

Pretending for a moment that we don't realise that things are going to get messy because "0.dev0" is a pre-release, the above is perfectly reasonable. However:

```
>>> Version("0.dev0") < Version("0")
True
```

So, of the two upper bounds, "<=0.dev0" is the tighter. However:

```
>>> Version("0.dev0") in SpecifierSet("<=0.dev0")
True
```

The problem here is that "<0" is actually a tighter bound in the sense that it excludes pre-release versions even if they are <0. And setting the `prereleases` flag on the specifier set doesn't address this (or at least, I can't prove to my satisfaction that it does) because the set would then behave differently (you can't combine it using `&` with a set that says `prereleases=True` for example).

**Comment by pfmoore** [comment]
> I don't remember thinking to hard about it, and just did the ones that explicitly included the pre-release that was being mentioned, which meant that > and < were excluded from being considered triggering the "user has requested a prerelease" logic... but I dont' feel really strongly about that one way or another.

Yeah, I'm pretty sure the whole "prereleases" logic (here, in the spec, and in pip) was all based on an instinctive desire to have things "do what I mean", but failed to create something that was mathematically or logically consistent. Regrettably, though, I think we're probably stuck with what we have, short of a "Version specifiers 2.0" like you mentioned on Discourse...

**Comment by dstufft** [comment]
Yea, in hindsight the logic probably should have been either simplified or pushed to the edges (e.g. in pip/poetry/etc), and left the core library and spec alone to implement the logically consistent thing.

**Comment by pfmoore** [comment]
I guess a fair question for the `packaging` maintainers is whether they (you) would support a breaking change that did just that - made prereleases just like any other version, and left it to tools to implement the "do what I mean" semantics.

To be clear, I'm not even sure that's possible, but unless the maintainers have the stomach for that level of breakage, it's not worth putting effort into trying to find a workable approach. (Personally, I expect the answer to be "no, this is too much for too little benefit" - I know I wouldn't be happy trying to deal with the knock-on effects of such a change in pip).

**Comment by dstufft** [comment]
Brett and Pradyun's opinions should weigh much higher than mine, since they've been the primary maintainers for awhile now.

Personally I think it's something we should do, but I have a bit of a preference for coupling it to the hypothetical PEP 440v2, but if someone felt strongly about getting that particular change out and figured out a good way to manage the breakage, I wouldn't (personally) be super upset about that.

But again, I'd consider my opinion to be less of a maintainer of packaging at this point and more just a guy whose been around for awhile :)

**Comment by notatallshaw** [comment]
> The problem here is that "<0" is actually a tighter bound in the sense that it excludes pre-release versions even if they are <0. And setting the `prereleases` flag on the specifier set doesn't address this (or at least, I can't prove to my satisfaction that it does) because the set would then behave differently (you can't combine it using `&` with a set that says `prereleases=True` for example).

Yes, I arrived at similar concerns through different means. For example, I was worried that for something like ">0,<0," there might actually be versions that would pass this Specifier Set if one of them is a pre-release (it seems from this discussion that's not the case in this example).

I still think it's possible to determine in many cases if a Specifier Set is contradictory, e.g., ">2,<1". However, in this case, it doesn't matter if one of them is a pre-release. The boundary conditions are particularly difficult, exponentially so because of pre-releases.

I had an idea of writing a testing mechanism that generates all possible versions around a boundary version (e.g., 1.0), up to some topological equivalence. Then, it creates all possible Specifiers around that boundary version and checks what versions are contained in that specifier. This can be used to test various things, like if Specifier Sets are logically consistent with their underlying Specifiers, and to test a function that determines if a Specifier Set is completely exclusionary or not. I have this mechanism mostly working, but there were some nuances I hadn't ironed out which this thread, I think, should immensely help with when I have a moment to work on it again.

**Comment by notatallshaw** [comment]
FYI I've made a PR (https://github.com/pypa/packaging/pull/794) to follow the spec, which will also fix this imbalance.

**Comment by notatallshaw** [comment]
FYI, there is a package which already does this: https://github.com/pdm-project/dep-logic

I have a [branch](https://github.com/pypa/pip/compare/main...notatallshaw:pip:speedy-resolve) that integrate's it into resolvelib / pip's resolution process, and it seems to work quite well.

**Comment by notatallshaw** [comment]
> The problem here is that "<0" is actually a tighter bound in the sense that it excludes pre-release versions even if they are <0. And setting the `prereleases` flag on the specifier set doesn't address this (or at least, I can't prove to my satisfaction that it does) because the set would then behave differently (you can't combine it using `&` with a set that says `prereleases=True` for example).

I've been thinking about this a bit recently as the ability for a specifier set to have a distinct canonical form is a prerequisite of more advanced resolution algorithms.

As you're aware, trying to map the Python version specifier `<V` to the mathematical inequality relation `<` does not hold, because the Python version specifier `<V` does not preserve ordinality of V and is not transitive.

The versions `V` themselves are ordinal under inequality:

```python
>>> Version("1.0a0") < Version("1.0b0")
True
>>> Version("1.0b0") < Version("1.0")
True
```

But the Specifier `<V` does not preserve this:

```python
>>> bool(list(Specifier("<1.0b0").filter(["1.0a0"])))
True
>>>  bool(list(Specifier("<1.0").filter(["1.0b0"])))
False
```

And this is not related to pre-release vs. not pre-release:

```python
>>> bool(list(Specifier("<2.0").filter(["1.0b0"])))
True
```

It is related to the line in the spec:

> The exclusive ordered comparison <V MUST NOT allow a pre-release of the specified version unless the specified version is itself a pre-release. 

One way I've thought to handle this, is when simplifying any SpecifierSet where there is a specifier with final version V, `<V` is to convert it to `<Vdev0,prerelease=False`, this version has the same semantics as `<V` but the ordinality of V is preserved.

There are however bugs that need fixing first, as packaging does not consistently implement the semantics of prereleases across the various methods of Specifier and SpecifierSet, e.g. https://github.com/pypa/packaging/issues/856 and https://github.com/pypa/packaging/issues/895

**Comment by brettcannon** [comment]
> Brett and Pradyun's opinions should weigh much higher than mine, since they've been the primary maintainers for awhile now.

I'm fine with such functionality.

**Comment by notatallshaw** [comment]
While simplifying a `SpecifierSet` may, or may not, be possible , what I do think is possible is determining the unsatisfiability of a `SpecifierSet`, and this is good enough step for a lot of possible optimizations in a resolver. I have therefore raise a new issue https://github.com/pypa/packaging/issues/940 and plan to raise a PR to implement that shortly.

**Comment by notatallshaw** [comment]
I have an update to this, I'm going to add a public API for pubgrub-style version ranges (https://github.com/pypa/packaging/issues/1143), currently my idea is that you will be able to do `SpecifierSet.from_ranges(specifier_set.to_ranges())` and this will implicitly get you a simplified version of a `SpecifierSet`, but won't necessarily guarantee any particular form or style of simplification.

I've done the initial work of adding some range mechanisms in https://github.com/pypa/packaging/pull/1119 (which working on revealed several specifier set edge cases not correctly handled by packaging), next I'm going to work on https://github.com/pypa/packaging/pull/1120 and a Public API PR on top of that.

**Comment by pfmoore** [comment]
I'm a little concerned about this - will these ranges map cleantly to PEP 440 version specifiers? If not, then it seems to me that there would be a significant potential for confusion where people get semantics that are subtly different from what they expect. While I can see the attraction of working with a cleaner model for specifiers, we still need to conform to Python's standards. And if a better model requires changing the standards, we can do that[^1].

If the proposal is for something that will remain fully PEP 440 compliant, then sorry for the unnecessary noise 🙂

[^1]: If changing the standards is the sticking point here, that's a problem with our standards process, not a reason for tools (especially a library as foundational as `packaging`) to avoid following them.

**Comment by notatallshaw** [comment]
> will these ranges map cleantly to PEP 440 version specifiers?

Yes, in fact thinking on it one approach may be to extend `SpecifierSet` to support all the required methods for implementing pubgrub-style algorithms, such a negation. I have a fast and working pubgrub implementation that can resolve Python packages, it currently uses a fork of packaging that implements version ranges, I'll write out extending `SpecifierSet` as well and see what the trade offs are in the different designs, I suspect there will be some downsides to overloading `SpecifierSet`s with two different concerns, but I can try it and see.

> If not, then it seems to me that there would be a significant potential for confusion where people get semantics that are subtly different from what they expect. While I can see the attraction of working with a cleaner model for specifiers, we still need to conform to Python's standards.

Fully agreed, however the main confusion comes from the standard itself, because the PEP 440 version specifiers are described with semantics that are not met by PEP 440 versions.

You can see this with the new `_BoundryVersion` class I have implemented to to support `is_unsatisfiable()`:  https://github.com/pypa/packaging/blob/main/src/packaging/specifiers.py#L61. This kind of "implied by the spec but not exposed by the spec" object would remain an implementation detail not in the public API.

Likewise uv's extension to the PEP 440 version adds a "max" boolean that will always be sorted above versions that have the same segments: https://github.com/astral-sh/uv/blob/main/crates/uv-pep440/src/version.rs#L1727, and then has supporting functions `with_min` and `with_max` to get PEP 440 version specifier ordering correct: [https://github.com/astral-sh/uv/blob/main/crates/uv-pep440/src/version.rs#L764 / ](https://github.com/astral-sh/uv/blob/main/crates/uv-pep440/src/version.rs#L757-L796)

Whereas Poetry's `VersionRange` object is more simple, it only includes the max and min version, and whether they are open bounded or close bounded: https://github.com/python-poetry/poetry-core/blob/main/src/poetry/core/constraints/version/version_range.py#L23, but then they have to spread out PEP 440 semantics in different parts of the finder/provider stack: https://github.com/python-poetry/poetry/blob/main/src/poetry/repositories/repository.py#L36 and https://github.com/python-poetry/poetry/blob/main/src/poetry/puzzle/provider.py#L305.

My aim is that if you are writing a Python packaging resolver you should be able to rely on packaging itself to implement all the PEP 440 semantics and not have to modify the objects or implement the semantics yourself to comply with the specification.

> And if a better model requires changing the standards, we can do that[1](https://github.com/pypa/packaging/issues/776#user-content-fn-1-a67fc61b7954aa7a3cb3f492fa70b023).

The standards are *just* consistent enough to implement a modern resolver algorithm on top of, but they require a lot of edge case catching which historically hasn't been done by `packaging` because it's never had to worry about mathematical consistency. 

My plan is to keep adding more useful property based testing to validate mathematical consistent rules that are implied by the spec or implied by the APIs we provide: https://github.com/pypa/packaging/tree/main/tests/property

**Comment by notatallshaw** [comment]
I have a draft PR https://github.com/pypa/packaging/pull/1182  (that depends on https://github.com/pypa/packaging/pull/1120) which implements round tripping normalization, which effectively simplifies:

```python
>>> str(SpecifierSet("!=1.0,>2,>=3").to_range().to_specifier_set())
'>=3'
```

**Comment by henryiii** [comment]
That's in now, so can this be closed?

**Comment by pfmoore** [comment]
I think so, yes.

**Comment by notatallshaw** [comment]
I am thinking about adding an explicit method on SpecifierSet that can leverage the information of the specifiers it is coming from to always guarantee same or simpler, but I have no timeline for that.

**Comment by henryiii** [comment]
I could imagine a `.simplify()` that internally basically calls `str(SpecifierSet("!=1.0,>2,>=3").to_range().to_specifier_set())`. That one might want to use or at least keep `~=`, though, so maybe it's a little more complex than that. (Or we could just define `~=` as always being not simpler. :) )
